File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void solve(vector<int>& nums, vector<vector<int>> &ans, int l, int r){
        if(l == r){
            ans.push_back(nums);
            return;
        }
        for(int i = l; i <= r; i++){
            swap(nums[i], nums[l]);
            solve(nums, ans, l+1, r);
            swap(nums[i], nums[l]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        solve(nums, ans, 0, nums.size()-1);
        return ans;
        
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        boolean[] used = new boolean[nums.length];
        Arrays.fill(used, false);

        solve(nums, 0, new ArrayList<>(), ans, used);
        return ans;
    }

    public void solve(int[] nums, int start, List<Integer> temp, List<List<Integer>> ans, boolean[] used){

        if(temp.size() == nums.length){
            ans.add(new ArrayList<>(temp));
            return;
        }

        for(int i = 0; i < nums.length; i++){

            if(used[i] == true)
                continue;

            temp.add(nums[i]);
            used[i] = true;

            solve(nums, i, temp, ans, used);

            temp.remove(temp.size() - 1);
            used[i] = false;
        }
    }
}